src/main/java/org/wildfly/security/auth/realm/token/validator/OAuth2IntrospectValidator.java

```diff
@@ -121,98 +109,24 @@ private boolean isValidToken(JsonObject claims) {
-            parameters.put("token_type_hint", "access_token");
-
-            byte[] params = buildParameters(parameters);
-
-            connection.setDoOutput(true);
-            connection.setRequestMethod("POST");
-            connection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
-            connection.setRequestProperty("Content-Length", String.valueOf(params.length));
-            connection.setRequestProperty("Authorization", "Basic " + CodePointIterator.ofString(clientId + ":" + clientSecret).asUtf8().base64Encode().drainToString());
-
-            try (OutputStream outputStream = connection.getOutputStream()) {
-                outputStream.write(params);
-            }
-
-            try (InputStream inputStream = new BufferedInputStream(connection.getInputStream())) {
-                return Json.createReader(inputStream).readObject();
-            }
-        } catch (IOException ioe) {
-            if (connection != null && connection.getErrorStream() != null) {
-                InputStream errorStream = connection.getErrorStream();
-
-                try (BufferedReader reader = new BufferedReader(new InputStreamReader(errorStream))) {
-                    StringBuffer response = reader.lines().reduce(new StringBuffer(), StringBuffer::append, (buffer1, buffer2) -> buffer1);
-                    log.errorf(ioe, "Unexpected response from token introspection endpoint [%s]. Response: [%s]", tokenIntrospectionUrl, response);
-                } catch (IOException e) {
-                    throw log.tokenRealmOAuth2TokenIntrospectionFailed(ioe);
-                }
-            } else {
-                throw log.tokenRealmOAuth2TokenIntrospectionFailed(ioe);
-            }
-        } catch (Exception e) {
-            throw log.tokenRealmOAuth2TokenIntrospectionFailed(e);
-        }
-
-        return null;
-    }
-
-    private HttpURLConnection openConnection(URL url, SSLContext sslContext, HostnameVerifier hostnameVerifier) throws IOException {
-        Assert.checkNotNullParam("url", url);
-
-        boolean isHttps = url.getProtocol().equalsIgnoreCase("https");
-
-        if (isHttps) {
-            if (sslContext == null) {
-                throw log.tokenRealmOAuth2SSLContextNotSpecified(url);
-            }
-
-            if (hostnameVerifier == null) {
-                throw log.tokenRealmOAuth2HostnameVerifierNotSpecified(url);
-            }
+            return HttpClient.builder().sslContext(sslContext).hostNameVerifier(hostnameVerifier).build()
+                    .post(tokenIntrospectionUrl)
+                        .header(HttpClient.HttpHeader.contentType("application/x-www-form-urlencoded"))
+                        .header(HttpClient.HttpHeader.authorizationBasic(clientId, clientSecret))
+                        .param("token", token)
+                        .param("token_type_hint", "access_token")
+                        .execute((inputStream, throwable) -> {
```

Comment:
Right now, it follows the redirect. Will that and avoid "follow redirects". We don't need to handle any 3xx status code from the server.